### DISCUSSION TITLE
[ChainSync client] More data regarding Tx inputs

### DISCUSSION THREAD

**Discussion by KristianBalaj** [CardanoSolutions/ogmios#304] [Q & A] [answered]
Currently the inputs for a Tx are of type `TxIn` containing just the `txId` and `index`.
It would be useful having more data about the inputs (address, value, datum, ...).

What is the reason why this data is not provided in the `rollForward` callback?
Is there any plan to include them?

Currently, the workaround is to separately query the UTXOs.

**Comment by KtorZ** [comment, marked as answer]
Ogmios provides an interface on the base protocols of the Cardano node. The chain sync protocol provides access to raw blocks. In blocks, transaction inputs are stored as simple references (transaction id & output index). Hence why you get them in this form for Ogmios too.

Being able to resolve address, value, datum and other elements associated with an input requires indexing. The node (and thus Ogmios) provides an API to query unspent outputs by reference through the state query protocol. This is because the ledger does maintain a UTxO set of the entire chain. However, you can't resolve spent inputs with that.


if you need more elaborate indexing, e.g. the ability to query any past transactions or, a way to query information other than by reference, you may want to look at [kupo](https://github.com/cardanosolutions/kupo) which is a bespoke tool for the job.

**Reply by KristianBalaj** [reply]
Thanks 🙌
